File: algorithms/11_golang/numerical/rayleigh_quotient.go
package main

import (
	"fmt"
	"math"
)

func rayleighQuotient(matrix [][]float64, vector []float64) float64 {
	n := len(matrix)
	scale := 0.0
	for _, value := range vector {
		scale = math.Max(scale, math.Abs(value))
	}
	x := make([]float64, n)
	for i := range vector {
		x[i] = vector[i] / scale
	}
	eigenvalue := 0.0
	for iteration := 0; iteration < 100; iteration++ {
		product := make([]float64, n)
		norm := 0.0
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				product[i] += matrix[i][j] * x[j]
			}
			norm = math.Max(norm, math.Abs(product[i]))
		}
		for i := 0; i < n; i++ {
			x[i] = product[i] / norm
		}
		numerator, denominator := 0.0, 0.0
		for i := 0; i < n; i++ {
			ax := 0.0
			for j := 0; j < n; j++ {
				ax += matrix[i][j] * x[j]
			}
			numerator += x[i] * ax
			denominator += x[i] * x[i]
		}
		next := numerator / denominator
		if math.Abs(next-eigenvalue) < 1e-12 {
			return next
		}
		eigenvalue = next
	}
	return eigenvalue
}

func main() {
	eigenvalue := rayleighQuotient([][]float64{{2, 1}, {1, 2}}, []float64{1, 0})
	if math.Abs(eigenvalue-3) > 1e-9 {
		panic("rayleigh failed")
	}
	fmt.Println("rayleigh quotient ok")
}

File: algorithms/07_typescript/math/rayleigh_quotient.ts
// Rayleigh quotient iteration.
function rayleighQuotient(matrix: number[][], vector: number[], iterations = 100): { eigenvalue: number; vector: number[] } {
  const n = matrix.length;
  let x = vector.map((value) => value / Math.max(...vector.map((v) => Math.abs(v))));
  let eigenvalue = 0;
  for (let iteration = 0; iteration < iterations; iteration += 1) {
    const product = matrix.map((row) => row.reduce((acc, value, j) => acc + value * x[j], 0));
    const norm = Math.max(...product.map((value) => Math.abs(value)));
    x = product.map((value) => value / norm);
    const numerator = x.reduce((acc, value, i) => acc + value * matrix[i].reduce((s, m, j) => s + m * x[j], 0), 0);
    const denominator = x.reduce((acc, value) => acc + value * value, 0);
    const next = numerator / denominator;
    if (Math.abs(next - eigenvalue) < 1e-12) return { eigenvalue: next, vector: x };
    eigenvalue = next;
  }
  return { eigenvalue, vector: x };
}

const { eigenvalue } = rayleighQuotient([[2, 1], [1, 2]], [1, 0]);
if (Math.abs(eigenvalue - 3) > 1e-9) throw new Error("rayleigh failed");
console.log(`eigenvalue=${eigenvalue}`);
